processor-api/src/main/java/com/synopsys/integration/alert/processor/api/filter/JobNotificationMapper.java

```diff
@@ -45,16 +46,31 @@ public JobNotificationMapper(JobAccessor jobAccessor) {
      * @param frequencies      an Additional filter to specify when querying data from the DB
-     * @return a {@code Map} where the distribution job is used to map to a list of notifications that were passed in.
+     * @return a {@code StatefulAlertPagedModel} where a page of distributions job is used to map to a list of notifications that were passed in.
      */
-    public List<FilteredJobNotificationWrapper> mapJobsToNotifications(List<DetailedNotificationContent> detailedContents, Collection<FrequencyType> frequencies) {
+    public StatefulAlertPagedModel<FilteredJobNotificationWrapper> mapJobsToNotifications(List<DetailedNotificationContent> detailedContents, List<FrequencyType> frequencies) {
+        AlertPagedModel<FilteredJobNotificationWrapper> pageOfJobsToNotification = mapPageOfJobsToNotification(detailedContents, frequencies, INITIAL_PAGE_NUMBER, PAGE_SIZE);
+        return new StatefulAlertPagedModel<>(
+            pageOfJobsToNotification.getTotalPages(),
+            pageOfJobsToNotification.getCurrentPage(),
+            pageOfJobsToNotification.getPageSize(),
+            pageOfJobsToNotification.getModels(),
+            (number, size) -> mapPageOfJobsToNotification(detailedContents, frequencies, number, size)
+        );
+    }
+
+    private AlertPagedModel<FilteredJobNotificationWrapper> mapPageOfJobsToNotification(List<DetailedNotificationContent> detailedContents, List<FrequencyType> frequencies, int pageNumber, int pageSize) {
         Map<FilteredDistributionJobResponseModel, List<NotificationContentWrapper>> groupedFilterableNotifications = new HashMap<>();
 
+        FilteredDistributionJobRequestModel filteredDistributionJobRequestModel = createRequestModelFromNotifications(detailedContents, frequencies);
+
+        AlertPagedModel<FilteredDistributionJobResponseModel> jobs = processingJobAccessor.getMatchingEnabledJobsByFilteredNotifications(filteredDistributionJobRequestModel, pageNumber, pageSize);
         for (DetailedNotificationContent detailedNotificationContent : detailedContents) {
-            List<FilteredDistributionJobResponseModel> filteredDistributionJobResponseModels = retrieveMatchingJobs(detailedNotificationContent, frequencies);
-            for (FilteredDistributionJobResponseModel filteredDistributionJobResponseModel : filteredDistributionJobResponseModels) {
-                List<NotificationContentWrapper> applicableNotifications = groupedFilterableNotifications.computeIfAbsent(filteredDistributionJobResponseModel, ignoredKey -> new LinkedList<>());
-                applicableNotifications.add(detailedNotificationContent.getNotificationContentWrapper());
+            for (FilteredDistributionJobResponseModel filteredDistributionJobResponseModel : jobs.getModels()) {
+                if (JobNotificationFilterUtils.doesNotificationApplyToJob(filteredDistributionJobResponseModel, detailedNotificationContent)) {
```

Comment:
Do we need this check? Shouldn't the jobs already be filtered by the notification content? What kind of case will return false with the given notificationContent?

Reply:
Yes this check is required because it will be using the JobNotificationFilterUtils to determine if the job applies to the notification in the list of notifications that it's looping through. When we get the jobs from line 67 we are getting a page of ALL jobs that match our pre-filtered list, but not all of these jobs match to every notification. We then need to loop through each notification and then through each job and run this check to determine if the job matches the notification. Once we have this we run the compileIfAbsent to add it to the groupedFilterableNotifications where the job is then mapped to all of its notifications. 

As an example, imagine you have 2 jobs, one for a HIGH severity vuln and another for a LOW severity vuln, and then 2 notifications for a HIGH and another for a LOW vuln. severity are produced. On line 67 we get a page containing 2 jobs, one HIGH and one LOW. Then we need to loop through the notifications but based on these jobs only the relevant notification gets mapped to each job.